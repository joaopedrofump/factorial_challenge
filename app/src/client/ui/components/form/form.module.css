.title {
  color: var(--secondary-color);
  font-size: 1rem;
  font-style: italic;
}

.form {
  background-color: var(--paper);
  display: flex;
  justify-content: space-between;
  padding: 1rem;
  border-radius: 7px;
}

.paper {
  box-shadow: 0 3px 10px rgb(0 0 0 / 0.2);
}

.form > input[type="text"],
.form > input[type="number"] {
  min-height: 2rem;
  border: none;
  border-bottom: 1px solid var(--gray);
  padding: 0.5rem;
  flex: 1;
  margin-right: 2rem;
}

.form > input[type="text"]:focus,
.form > input[type="number"]:focus {
  outline: none;
  border-bottom: 1.5px solid var(--primary-color);
}

.form > input[type="submit"] {
  border: none;
  border-radius: 5px;
  background-color: var(--secondary-color);
  color: #fff;
  width: 75px;
}

.form > input[type="submit"]:hover {
  opacity: 0.9;
}

@media (max-width: 600px) {

  .container {
    width: 100%;
  }

  .form {
    flex-direction: column;
  }

  .form > input {
    width: 100%;
  }

  .form > input[type="submit"] {
    width: 100%;
    height: 40px;
  }
  .form > input[type="text"],
  .form > input[type="number"] {
    margin-right: 0;
    margin-bottom: 1.5rem;
  }
}
