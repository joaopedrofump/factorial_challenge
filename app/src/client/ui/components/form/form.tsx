import React, { FC, useRef } from "react";
import { getMeasure, saveMeasure } from "src/client/service/measure-service";
import { Measurement } from "src/common/model/measure";

import styles from "./form.module.css";

type Props = {
  updateMeasures: (measurement: Measurement) => void;
  hcMeasurement?: string;
  title?: string;
  valuePlaceHolder?: string;
  elevated?: boolean;
};

const Form: FC<Props> = ({
  updateMeasures,
  hcMeasurement,
  title,
  valuePlaceHolder,
  elevated,
}) => {
  const measurementRef = useRef<HTMLInputElement | null>(null);
  const valueRef = useRef<HTMLInputElement | null>(null);

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const measurement = measurementRef?.current?.value as string;
    const value = valueRef?.current?.value as string;

    await saveMeasure(measurement, parseFloat(value));
    const updatedMeasurement = await getMeasure(measurement);

    updateMeasures(updatedMeasurement);
    (valueRef.current as HTMLInputElement).value = "";
  };

  return (
    <section className={styles.container}>
      {title && <h2 className={styles.title}>{title}</h2>}
      <form
        onSubmit={handleSubmit}
        className={`${styles.form} ${elevated ? styles.paper : ""}`}
      >
        <input
          placeholder="metric*"
          required
          hidden={typeof hcMeasurement !== "undefined"}
          ref={measurementRef}
          type="text"
          {...(typeof hcMeasurement !== "undefined" && {
            readOnly: true,
            value: hcMeasurement,
          })}
        ></input>
        <input
          step={0.1}
          placeholder={valuePlaceHolder ?? "value*"}
          required
          ref={valueRef}
          type="number"
        ></input>
        <input type="submit" value="save" />
      </form>
    </section>
  );
};

export { Form };
