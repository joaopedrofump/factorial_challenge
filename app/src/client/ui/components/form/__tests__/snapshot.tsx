import { render } from "@testing-library/react";
import { Form } from "src/client/ui/components/form/form";

jest.mock("src/client/ui/components/charts/charts", () => ({
  Charts: jest.fn(),
}));

describe("test several shapes of the form", () => {
  describe("when passing default measure", () => {
    it("renders form without name input", () => {
      const { container } = render(
        <Form updateMeasures={jest.fn()} hcMeasurement="hc" key={1} />
      );
      expect(container).toMatchSnapshot();
    });
  });
  describe("when not passing default measure", () => {
    it("renders form without name input", () => {
      const { container } = render(
        <Form updateMeasures={jest.fn()} elevated title="New meeasurment" key={1} />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
