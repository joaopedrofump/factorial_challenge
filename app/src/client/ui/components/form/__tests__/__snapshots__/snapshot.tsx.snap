// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test several shapes of the form when not passing default measure renders form without name input 1`] = `
<div>
  <section
    class="container"
  >
    <h2
      class="title"
    >
      New meeasurment
    </h2>
    <form
      class="form paper"
    >
      <input
        placeholder="metric*"
        required=""
        type="text"
      />
      <input
        placeholder="value*"
        required=""
        step="0.1"
        type="number"
      />
      <input
        type="submit"
        value="save"
      />
    </form>
  </section>
</div>
`;

exports[`test several shapes of the form when passing default measure renders form without name input 1`] = `
<div>
  <section
    class="container"
  >
    <form
      class="form "
    >
      <input
        hidden=""
        placeholder="metric*"
        readonly=""
        required=""
        type="text"
        value="hc"
      />
      <input
        placeholder="value*"
        required=""
        step="0.1"
        type="number"
      />
      <input
        type="submit"
        value="save"
      />
    </form>
  </section>
</div>
`;
